src/Elmish.WPF/ViewModel.fs

```diff
@@ -684,6 +696,11 @@ and [<AllowNullLiteral>] internal ViewModel<'model, 'msg>
           |> ValueOption.map (fun vm -> b.SubModelSeqBinding.GetId vm.CurrentModel)
         b.Set id model
         true
+    | Cached b ->
+        let successful = trySetMember model value b.Binding
+        if successful then
+          b.Cache := None // `None` is safter, but consider using `Some value` instead
```

Comment:
This is in `trySetMember`.  If the set was successful, then `true` is returned and the cache is stale.  We can either assign `None` to empty the cache or assign `Some value` to populate the cache with the value that was just set.

The behavior of the caching currently in `master` for `OneWayLazy` and `SubModelSelectedItem` equivalent to assigning `None` here.  Assigning `Some value` instead would be a further optimization.

The risk with setting `Some value` is if anyone would purposefully return a different value than just set or if the user has a bug in their code that causes a different value to be returned than just set.  This "improved" caching would "hide" that bug--make it difficult to identify.

The Elm architecture is only supposed to store state in the model.  If we cache `Some value` here, then it might not be the same data that would be returned from `tryGetMember`.  This would definitely happen to someone eventually, and I don't expect they like that we made that choice.

So, assigning `None` mean the cache will be populated from the model and user-provided functions, which keeps things very understandable and predictable.  This is what I would prefer here.

Reply:
Thanks for the detailed explanation. Let's definitely keep `None`. There are valid use-cases for returning different data than what is being set (e.g. format-as-you-type for mobile numbers).

Reply:
Excellent example!

And that reminds me, I forgot to say this..

For each of the `OneWay`, `TwoWay`, and `TwoWayValidation` bindings, it is now a one-line change to add caching.  The example you gave for auto-formatting user input (like, say, making the first character capital if it is a letter) requires two-way functionality.

I suggest that we add this caching to `OneWay`, `TwoWay`, and `TwoWayValidation`.  Then I can document using a test that assigning `None` is the behavior that we want (i.e. the test will fail if `Some value` were assigned instead).

How does that sound?

Reply:
Hmm... I'm not sure if we should make all those bindings caching by default. There's some overhead involved in caching, and for trivial getters, non-cached bindings may be more performant. I have no idea how much overhead there is, though. As far as I know, it may improve performance, or it may not.

Thoughts?

Reply:
Your hesitation to add caching to the other three binding types is good.  We should to see how the performance compares.  I will open a separate issue for that.

I could make such a test for `SubModelSelectedItem`, but I think it would only be a pathological one.  One advantage of writing that test for, say, a `TwoWayBinding` is that I could use the excellent example behavior that you suggested.

Reply:
> Your hesitation to add caching to the other three binding types is good. We should test to see how the performance compares. I will open a separate issue for that.

Just did that with #184.

> Regarding the None test: Definitely 👍

How about we proceed without a test for now but I make an issue about adding such a test?

Reply:
> How about we proceed without a test for now but I make an issue about adding such a test?

Sure!

Reply:
Created issue #185